### PR TITLE
[Feature]file guid unification

### PR DESCRIPTION
**Pull request by chrisseto**
Code is read for review, a couple more unit tests need to get written.
See commit messages for details about changes.

Migration steps:
```bash
python -m scripts.osfstorage.migrate_to_generic warning
python -m scripts.migrate_file_guids
```

### CODE REVIEW

**Comment by sloria** [comment]
Note: This is a dependency for some of the work on API v2

File: framework/guid/model.py [hunk 1]
```diff
@@ -20,6 +20,26 @@ class Guid(StoredObject):
     _id = fields.StringField(primary=True)
     referent = fields.AbstractForeignField()
 
+    @classmethod
+    def generate(self, referent=None):
+        while True:
+            # Create GUID
+            guid_id = ''.join(random.sample(ALPHABET, 5))
```
Comment (lyndsysimon) [inline framework/guid/model.py]: Have we considered using `random.SysRandom` instead of `random`?

While our GUIDs aren't being used for security purposes, using the Python implementation can generate overlapping sequences of values in some threaded contexts.

From [the Python docs](https://docs.python.org/2/library/random.html):

> The functions supplied by this module are actually bound methods of a hidden instance of the random.Random class. You can instantiate your own instances of Random to get generators that don’t share state. This is especially useful for multi-threaded programs, creating a different instance of Random for each thread, and using the jumpahead() method to make it likely that the generated sequences seen by each thread don’t overlap.

I don't think this is something that needs changed right now, but it could impact performance when attempting to generate GUIDs concurrently.

**Comment by lyndsysimon** [comment]
Could you add a docstring to the scripts you've provided? It could be as simple as "Migrates OsfFile objects from ObjectIds to GUIDs" - just something so we'll know what they're for when flipping through in PyCharm.

File: tests/test_addons.py [hunk 2]
```diff
@@ -353,29 +355,29 @@ def setUpClass(cls):
         set_up_storage([DummyGuidFile], storage.MongoStorage)
 
 
-class TestAddonFileViewHelpers(OsfFileTestCase):
+# class TestAddonFileViewHelpers(OsfFileTestCase):
```
Comment (lyndsysimon) [inline tests/test_addons.py]: If this is no longer necessary, let's remove it instead of commenting it out. Otherwise, please add a comment indicating why it was commented out.

File: website/addons/osfstorage/tests/test_models.py [hunk 3]
```diff
@@ -15,61 +14,61 @@
...
-        mock_get.return_value = mock_response
-        mock_response.json.return_value = {
-            'data': {
-                'name': 'Morty',
-                'extra': {
-                    'version': 'Terran it up'
-                }
-            }
-        }
+# class TestFileGuid(OsfTestCase):
```
Comment (lyndsysimon) [inline website/addons/osfstorage/tests/test_models.py]: Lots of commented-out tests in this file - as I suggested above, let's either go ahead and remove them, or comment as to why they're commented out.

File: framework/guid/model.py (hunk 1 above)
Comment (sloria) [inline framework/guid/model.py]: Note: we do have a utility in `website.security` that uses `SystemRandom`.

File: website/addons/osfstorage/model.py [hunk 4]
```diff
@@ -28,17 +30,15 @@ class OsfStorageNodeSettings(StorageAddonBase, AddonNodeSettingsBase):
...
-    # Temporary field to mark that a record has been migrated by the
-    # migrate_from_oldels scripts
-    _migrated_from_old_models = fields.BooleanField(default=False)
+    root_node = fields.ForeignField('StoredFileNode')
 
     @property
     def folder_name(self):
         return self.root_node.name
 
+    def get_root(self):
```
Comment (lyndsysimon) [inline website/addons/osfstorage/model.py]: I think I understand what's going on and why, but this seems fragile. If a developer calls `<OsfStorageNodeSettings instance>.root_node`, they get a different object than if they call `<OsfStorageNodeSettings instance>.get_root()`.

I'm OK with it right now, I just wanted to poke at it.

**Comment by lyndsysimon** [comment]
First-pass code review complete. High-level functional review complete. I did not test the migration scripts.

File: website/addons/osfstorage/tests/test_models.py (hunk 3 above)
Comment (chrisseto) [inline website/addons/osfstorage/tests/test_models.py]: These are removed in https://github.com/CenterForOpenScience/osf.io/pull/3958

File: website/addons/osfstorage/model.py (hunk 4 above)
Comment (chrisseto) [inline website/addons/osfstorage/model.py]: There's not another option that seems great to me, developers need a way to get the wrapped variant of the root node and I'd prefer not to migrate what doesn't have to be migrated.

It doesn't feel super fragile to me; if you don't call `get_root()` nothing will work for you. :smiling_imp:

**Comment by chrisseto** [comment]
Quick update, I've encountered a small issue whilst writing tests. pause any review for the moment

**Comment by chrisseto** [comment]
@lyndsysimon Ready for round 2

**Comment by lyndsysimon** [comment]
Additional changes reviewed - looks good to me.

:star:

File: framework/guid/model.py (hunk 1 above)
Comment (sloria) [inline framework/guid/model.py]: @chrisseto Have you looked into this?

Comment (chrisseto) [inline framework/guid/model.py]: @sloria I have not, I personally think that's a bit out of scope of this PR. Nothing has changed about the amount of GUIDs being generated nor how they are being generated.
That being said I'd be happy to address it if you'd like me too.

Comment (sloria) [inline framework/guid/model.py]: Good point; I see now that this makes no changes to previous logic. This is fine as is.

File: scripts/osfstorage/migrate_to_generic.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    for node_settings in osfstorage_model.OsfStorageNodeSettings.find():
+        if node_settings.owner is None:
+            logger.warning('OsfStorageNodeSettings {} has no parent; skipping'.format(node_settings._id))
+            continue
+        logger.info('Migrating files for {}'.format(node_settings.owner.title))
+        root_node = osfstorage_model.OsfStorageFileNode.load(node_settings.to_storage()['root_node'])
+        if root_node is None:
+            logger.warning('OsfStorageNodeSettings {} has no root_node; skipping'.format(node_settings._id))
+            continue
+        list(osfstorage_model.OsfStorageNodeSettings.find(Q('node_settings', 'eq', node_settings._id)))
```
Comment (sloria) [inline scripts/osfstorage/migrate_to_generic.py]: Why is this line necessary?

Comment (sloria) [inline scripts/osfstorage/migrate_to_generic.py]: It seems incorrect; you're querying the `node_settings` field of the `OsfStorageNodeSettings` model.

File: website/files/models/base.py [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+    materialized_path = fields.StringField(required=True)
+
+    @property
+    def deep_url(self):
+        return self.wrapped().deep_url
+
+    def wrapped(self):
+        """Wrap self in a FileNode subclass
+        """
+        return PROVIDER_MAP[self.provider][int(self.is_file)](self)
```
Comment (sloria) [inline website/files/models/base.py]: Should a `KeyError` be handled here?

File: website/files/models/base.py [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+            if not create:
+                return None
+        return Guid.generate(self)
+
+
+class FileNodeMeta(type):
+    """Keeps track of subclasses of the ``FileNode`` object
+    Inserts all into the PROVIDER_MAP following the pattern:
+    {
+        provider: [ProviderFolder, ProviderFile, ProviderFileNode]
```
Comment (sloria) [inline website/files/models/base.py]: Why a list instead of a nested dict?

```python
{
    provider: {'folder_model': ProviderFolder, ...}
}
```

Comment (chrisseto) [inline website/files/models/base.py]: For cases such as [this](https://github.com/CenterForOpenScience/osf.io/pull/3940/files#diff-1d9389748ddcbd5854a7187ce37d1d66R81). A bool cast to an int is either 0 (Folder) or 1 (File).

Comment (sloria) [inline website/files/models/base.py]: Is that any better than:

```python
key = 'file_model' if self.is_file else 'folder_model'
return PROVIDER_MAP[self.provider][key](self)
```
?

File: website/files/models/base.py [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+            return cls.find_one(
+                Q('node', 'eq', node) &
+                Q('path', 'eq', path)
+            )
+        except NoResultsFound:
+            return cls.create(node=node, path=path)
+
+    @classmethod
+    def resolve_class(cls, provider, _type=2):
+        """Resolves a provider and type to the appropriate subclass.
```
Comment (sloria) [inline website/files/models/base.py]: Hate to nitpick, but PEP 257 recommends using the imperative ("Resolve a provider").

I dig the usage examples, though.

File: website/files/models/base.py (hunk 7 above)
Comment (sloria) [inline website/files/models/base.py]: Note: I don't feel strongly that it should be changed; just trying to understand the rationale.

File: website/files/models/base.py [hunk 9]
```diff
@@ -0,0 +1,588 @@
...
+        qs = filter(None, (qs,
+            Q('is_file', 'eq', cls.is_file) if hasattr(cls, 'is_file') else None,
+            Q('provider', 'eq', cls.provider) if hasattr(cls, 'provider') else None,
+        ))
+        # If out list is empty return None; there's no filters to be applied
+        if not qs:
+            return None
+        # Use reduce to & together all our queries. equavilent to:
+        # return q1 & q2 ... & qn
+        return reduce(lambda q1, q2: q1 & q2, qs)
```
Comment (sloria) [inline website/files/models/base.py]: Minor, but you can easily make this forwards-compatible by using `functools.reduce`.

File: website/files/models/base.py [hunk 10]
```diff
@@ -0,0 +1,588 @@
...
+
+    @classmethod
+    def find(cls, qs=None):
+        """A proxy for StoredFileNode.find but applies class based contraints.
+        Wraps The MongoQuerySet in a _GenWrapper this overrides the __iter__ of
+        MongoQuerySet to return wrapped objects
+        :rtype: _GenWrapper<MongoQuerySet<cls>>
+        """
+        return utils._GenWrapper(StoredFileNode.find(cls._filter(qs)))
```
Comment (sloria) [inline website/files/models/base.py]: `GenWrapper` seems to be a public class, so no need for the dangling underscore.

File: scripts/osfstorage/migrate_to_generic.py (hunk 5 above)
Comment (chrisseto) [inline scripts/osfstorage/migrate_to_generic.py]: Whoops..... nice catch, should have been `OsfStorageFileNode`

Comment (chrisseto) [inline scripts/osfstorage/migrate_to_generic.py]: Note: Its for pulling everything into odms cache

File: website/files/models/base.py (hunk 6 above)
Comment (chrisseto) [inline website/files/models/base.py]: I don't think so, if a key isn't found it should be an exception.

Comment (chrisseto) [inline website/files/models/base.py]: Or a custom exception that's a bit easier to understand

File: website/files/models/base.py (hunk 7 above)
Comment (chrisseto) [inline website/files/models/base.py]: The list forces a more binary format.
There is no way to have something in the system that is neither a file or folder without significant change.

File: website/files/models/base.py [hunk 11]
```diff
@@ -0,0 +1,588 @@
...
+from website.files import exceptions
+
+
+__all__ = (
+    'File',
+    'Folder',
+    'FileNode',
+    'FileVersion',
+    'StoredFileNode',
+    'TrashedFileNode',
```
Comment (sloria) [inline website/files/models/base.py]: :triangular_ruler:

File: website/files/models/base.py (hunk 6 above)
Comment (sloria) [inline website/files/models/base.py]: Yes, a custom exception would make sense.

File: website/files/models/base.py [hunk 12]
```diff
@@ -0,0 +1,588 @@
...
+    def generate_download_url(self, **kwargs):
+        return util.waterbutler_url_for(
+            'download',
+            self.provider,
+            self.path,
+            self.node,
+            **kwargs
+        )
+
+    def touch(self, revision=None, **kwargs):
```
Comment (sloria) [inline website/files/models/base.py]: It appears that this can also return a tuple. Can you document that behavior?

File: api/files/serializers.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        'id',
+        'size',
+        'identifier',
+        'content_type',
+    ])
+    id = ser.CharField(read_only=True, source='_id')
+    size = ser.IntegerField(help_text='The size of this file at this version')
+    identifier = ser.CharField(help_text='This version\'s unique identifier from it\'s original service')
+    content_type = ser.CharField(help_text='The mime type of this file at this verison')
+    links = LinksField({'html': 'absolute_url'})
```
Comment (brianjgeiger) [inline api/files/serializers.py]: I think it'll need a self link as well.

File: api/files/serializers.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    ])
+    id = ser.CharField(read_only=True, source='_id')
+    name = ser.CharField(help_text='Display name used in the general user interface')
+    path = ser.CharField(help_text='The unique path used to reference this object')
+    provider = ser.CharField(help_text='The Add-on service this file originates from')
+    last_touched = ser.DateTimeField(help_text='The last time this file had information fetched about it via the OSF')
+    # history = ser.ListField(ser.DictField(), help_text='A raw dump of the metadata recieved whenever infomation is fetched about it')
+
+    links = LinksField({
+        'html': 'absolute_url',
```
Comment (brianjgeiger) [inline api/files/serializers.py]: I think a self-link here as well.

File: api/files/permissions.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, FileNode), 'obj must be a Node or Pointer, got {}'.format(obj)
+        auth = get_user_auth(request)
+        if request.method in permissions.SAFE_METHODS:
+            return obj.node.is_public or obj.node.can_view(auth)
+        else:
+            if not obj.node.can_edit(auth):
+                return False
+
+            return any(
```
Comment (icereval) [inline api/files/permissions.py]: Why would you not want to use 'or' here?

File: website/util/__init__.py [hunk 16]
```diff
@@ -138,14 +140,19 @@ def waterbutler_url_for(route, provider, path, node, user=None, **kwargs):
         url.args['cookie'] = user.get_or_create_cookie()
     elif website_settings.COOKIE_NAME in request.cookies:
         url.args['cookie'] = request.cookies[website_settings.COOKIE_NAME]
+    elif 'Authorization' in request.headers:
+        url.args['token'] = request.headers['Authorization'].replace('Bearer ', '')
```
Comment (icereval) [inline website/util/__init__.py]: Authorization header could also be `Bearer ...` or `Basic ...` authentication. Might be better off checking `session.data.get('auth_user_access_token')`.

File: website/util/__init__.py [hunk 17]
```diff
@@ -117,7 +118,8 @@ def is_json_request():
 
 
 def waterbutler_url_for(route, provider, path, node, user=None, **kwargs):
-    """Reverse URL lookup for WaterButler routes
+    """DEPRECATED Use waterbutler_api_url_for
```
Comment (icereval) [inline website/util/__init__.py]: Will the cleanup PR remove this function?

File: api/files/permissions.py (hunk 15 above)
Comment (icereval) [inline api/files/permissions.py]: Why not use `or` here?

File: website/templates/project/view_file.mako [hunk 18]
```diff
@@ -196,6 +195,7 @@
           canEdit: ${ int(user['can_edit']) | sjson, n }
         }
       });
+    window.contextVars.file.urls.external = window.contextVars.file.extra.webView;
```
Comment (icereval) [inline website/templates/project/view_file.mako]: indent

File: website/views.py [hunk 19]
```diff
@@ -324,7 +324,7 @@ def _build_guid_url(base, suffix=None):
         each.strip('/') for each in [base, suffix]
         if each
     ])
-    return u'/{0}/'.format(url)
+    return u'/{0}/'.format(u''.join(url))
```
Comment (icereval) [inline website/views.py]: I know of the now famous unicode join hack, though please add a comment :pray:

File: website/util/__init__.py (hunk 16 above)
Comment (icereval) [inline website/util/__init__.py]: Hold on changing this to use `auth_user_access_token` (can still validate Bearer). Need to finish my evaluation of api v1.

File: api/files/permissions.py (hunk 15 above)
Comment (chrisseto) [inline api/files/permissions.py]: Seemed slightly more readable IMO

File: website/util/__init__.py (hunk 16 above)
Comment (chrisseto) [inline website/util/__init__.py]: This function should be removed anyways

File: website/util/__init__.py (hunk 17 above)
Comment (chrisseto) [inline website/util/__init__.py]: Yes it will.

File: api/files/permissions.py [hunk 20]
```diff
@@ -0,0 +1,20 @@
+from rest_framework import permissions
+
+from api.base.utils import get_user_auth
+from website.files.models import FileNode
+
+
+class ContributorOrPublic(permissions.BasePermission):
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, FileNode), 'obj must be a Node or Pointer, got {}'.format(obj)
```
Comment (sloria) [inline api/files/permissions.py]: The error message is incorrect.

File: api/files/views.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+        return FileNode.find(query)
+
+
+class FileDetail(generics.RetrieveUpdateAPIView, FileMixin):
+    """Details about a specific file.
+    """
+    serializer_class = FileSerializer
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        ContributorOrPublic,
```
Comment (sloria) [inline api/files/views.py]: This also needs `ReadOnlyIfRegistration`.

File: api/files/views.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+
+        obj = get_object_or_error(FileNode, key)
+
+        if check_permissions:
+            # May raise a permission denied
+            self.check_object_permissions(self.request, obj)
+        return obj
+
+
+class FileList(generics.ListAPIView, ODMFilterMixin):
```
Comment (sloria) [inline api/files/views.py]: There is no urlpattern for this view; is this intended?

File: api/files/serializers.py [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+
+class FileVersionSerializer(JSONAPISerializer):
+    filterable_fields = frozenset([
+        'id',
+        'size',
+        'identifier',
+        'content_type',
+    ])
+    id = ser.CharField(read_only=True, source='_id')
+    size = ser.IntegerField(help_text='The size of this file at this version')
```
Comment (sloria) [inline api/files/serializers.py]: Shouldn't these metadata be `read_only`?
